sdk/trace/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,215 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ratelimit // import "go.opentelemetry.io/otel/sdk/trace/ratelimit"
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+type config struct {
+	interval time.Duration
+	nowfunc  func() time.Time
+}
+
+type Option interface {
+	apply(*config)
+}
+
+type window struct {
+	// start is the beginning of this window
+	start time.Time
+
+	// low is the lower power-of-two probability
+	low trace.Sampler
+	// high is the higher power-of-two probability
+	high trace.Sampler
+
+	// lowProb is the probability of sampling at the lowProb
+	lowProb float64
+
+	count   int64
+	compute sync.Once
+}
+
+type Sampler struct {
```

Comment:
Suggestion: add a paragraph describing the idea of the algorithm in words, as it is much more difficult to follow in the code.

Reply:
Done, thanks.